### PR TITLE
Extend code_typed to be able to debug constant prop

### PR DESCRIPTION
**Pull request by Keno**
I'm currently debugging code that uses a lot of arrays with
dimensions in their type parameters. This type of code heavily
relies on constant propagation to lift dimensions from the value
domain into the type domain. Unfortunately, it's a bit hard to
discover what exactly causes inference to drop information from
constants, because there's no way to feed in constants for a
particular invocation using code_typed. This is a quick hack to
remidy that, by making `$`-interpolated expressions available
as constants to type inference, e.g.

```julia
julia> @code_typed 1+1
CodeInfo(
53 1 ─ %1 = (Base.add_int)(x, y)::Int64
   └──      return %1
) => Int64

julia> @code_typed $(1)+$(1)
CodeInfo(
53 1 ─     return 2
) => Int64
```

This is missing handling for a bunch of the other cases in code_typed, which is why this is WIP. I plan to put those in and finish this up once I've confirmed this is useful, but figured I'd put it up in the meantime.

### CODE REVIEW

**Comment by vtjnash** [comment]
This capability does seem useful. I'd like to base it on top of #28955 though, so that hopefully the internal API can be better.

**Comment by Keno** [comment]
Yes, I agree. This also needs to be cleaned up a bit, but I've been using it quite extensively and it's been very useful.

**Comment by Keno** [comment]
Rebased on top of current master (i.e. including #28955), extended to handle a few extra syntax forms. Additionally, the Symbol argument in get/setproperty! now gets automatically consted (only in the syntax form of course).

**Comment by Keno** [comment]
In any case, this is good to go from my perspective. Any further review comments? Otherwise I'll plan to merge in a day or so.

File: base/reflection.jl [hunk 1]
```diff
@@ -883,7 +888,16 @@ function code_typed(@nospecialize(f), @nospecialize(types=Tuple); optimize=true)
...
     for x in _methods(f, types, -1, world)
         meth = func_for_method_checked(x[3], types)
-        (code, ty) = Core.Compiler.typeinf_code(meth, x[1], x[2], optimize, params)
+        argtypes = nothing
+        if constvals !== nothing
+            argtypes = Vector(undef, length(types.parameters)+1)
+            for i = 1:length(types.parameters)+1
+                argtypes[i] = isassigned(constvals, i) ?
+                    Core.Compiler.Const(constvals[i]) : (
+                        i == 1 ? typeof(f) : types.parameters[i-1])
```
Comment (vtjnash) [inline base/reflection.jl]: There appear to be a couple issues with this (invalid assumptions about the structure of `types`, `typeof` instead of `Core.Typeof`). Would it be possible to format this better too? The deeply nested conditionals and indentation are hard to follow.

**Review by vtjnash — CHANGES_REQUESTED** [review]
I'm not sure I'll have a chance to finish review this week. Can you ping me again next week, unless someone else can make sure the `_gen_call_with_extracted_types` are correct and won't make it difficult to improve this code in the future (#20159 is the only open issue I can find to reference, so maybe we didn't open / prematurely closed the others)?

**Comment by Keno** [comment]
@vtjnash ping. Other than the obvious `Typeof`, could you suggest what the correct way to write this is?

**Comment by jrevels** [comment]
Bump. @vtjnash ping again 🙂

**Comment by vchuravy** [comment]
Bump. This would be quite useful to have for https://github.com/vchuravy/Cthulhu.jl!

**Comment by Matthew-D-Dwyer** [comment]
This issue is a few years old now, is it worth keeping it open or can it be closed? 

Matt.

**Comment by timholy** [comment]
Cthulhu is now quite good for debugging constprop, and that seems like a fine place for me. But I'll let @Keno decide whether to close this.

**Comment by vchuravy** [comment]
I think the question is do we want to expose entering type-inference with const-args on the initial frame?

```
code_typed((Float64,)) do (x,)
   sin(1.0) + x
end
```
Is the current "canonical" workaround.
cc: @aviatesk
